Recipient box typeahead for streams and topics

kaustubh-nair opened the pull request.

kaustubh-nair commented on zulipterminal/helper.py:
```diff
@@ -490,6 +490,11 @@ def match_emoji(emoji: str, text: str) -> bool:
     return emoji.lower().startswith(text.lower())
 
 
+def match_stream_names(stream_names: List[str], search_text: str) -> List[str]:
+    return [name for name in stream_names
+            if name.lower().startswith(search_text.lower())]
+
+
```
There are no pinned stream considerations here - I'm not using the existing function.
Refactoring `match_streams` looks a little hard, and pinning order is not super important - so I'd like to deal with this in a follow-up.

kaustubh-nair replied:
Though one hacky way to use the current `match_streams` would be to pass `[(stream_name, stream_name)..]` in data`.

kaustubh-nair replied:
They are certainly an improvement, but I don't see it as being a necessity for this particular feature. I'll make a temp fix but open an issue and make a follow-up PR for refactoring `match_stream`

neiljp commented:
@kaustubh-nair Is this PR supposed to be functional at this point?

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -63,9 +64,11 @@ def private_box_view(self, button: Any=None, email: str='') -> None:
+                        ) -> None:
         self.set_editor_mode()
         self.to_write_box = None
+        self.stream_id = stream_id
```
This gets set only on set-up, so if the stream is changed before the message is sent, the topic autocomplete uses the stream used at initialization. This is the same issue present with another PR, maybe?

kaustubh-nair replied:
Yes, I can fix it now that #738 is merged

neiljp replied:
This is resolved by the change at the end  of this commit?

kaustubh-nair replied:
Yes that should keep `stream_id` in sync

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -144,13 +144,13 @@ def generic_autocomplete(self, text: str, state: Optional[int]
         self.view.set_typeahead_footer(reduced_suggestions,
                                        state, is_truncated)
+        if typeahead:
+            typeahead = text[:prefix_index] + typeahead
```
Is this change necessary, given the next commit? Aren't you moving it back?

kaustubh-nair replied:
The git diff is a bit weird so it looks a little strange, but this statement stays in the same method.

kaustubh-nair commented on tests/ui_tools/test_boxes.py:
```diff
@@ -251,6 +251,39 @@ def test_generic_autocomplete_emojis(self, write_box, text,
+        'invalid_state-greater_than_possible_index',
+        'invalid_state-less_than_possible_index',
+    ])
+    def test__process_typeaheads(self, write_box, typeaheads, suggestions,
+                                 state, expected_state, expected_typeahead,
+                                 is_truncated, mocker):
+        write_box.view.set_typeahead_footer = mocker.patch(
+                                'zulipterminal.ui.View.set_typeahead_footer')
+
+        typeahead = write_box._process_typeaheads(typeaheads, state,
+                                                  suggestions)
+
+        assert typeahead == expected_typeahead
+        write_box.view.set_typeahead_footer.assert_called_once_with(
+                            suggestions[:10], expected_state, is_truncated)
+
```
I'm not sure what to do about the existing `generic_autocomplete` tests now, since now they span two different methods. Changing them to check for `_process_typeaheads` would be hard because the parameters would get complex. (Because we'd have to create lists of suggestions)

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -103,6 +108,19 @@ def stream_box_view(self, stream_id: int, caption: str='', title: str='',
+        streams = [stream[0]
+                   for stream in streams_list]
+
+        stream_data = list(zip(streams, streams))
```
Is this okay, or should `match_stream` be refactored?

kaustubh-nair replied:
Added

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -103,6 +108,19 @@ def stream_box_view(self, stream_id: int, caption: str='', title: str='',
+        matched_streams = match_stream(stream_data, text,
+                                       self.view.pinned_streams)
+
+        return self._process_typeaheads(matched_streams[0], state,
+                                        matched_streams[1])
```
`matched_stream[0]` and `matched_stream[1]` have the same data, but I've used both of them for the sake of consistency.

kaustubh-nair replied:
Added

kaustubh-nair commented:
@neiljp Thanks for the review! The problem you see for stream autocomplete is also the same bug caused due to readline's ability to autocomplete only words. It shows all possible suggestions for a an empty search string.
I've pushed a newer version with the following updates:
* Added tests with cases for failing whitespace errors.
* Fixed for changing topics by dynamically loading them when needed.
* New commit for improving invalid stream error message.

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -134,10 +133,19 @@ def generic_autocomplete(self, text: str, state: Optional[int]
+            typeahead = text[:prefix_index] + typeahead
+        return typeahead
+
+    def _process_typeaheads(self, typeaheads: List[str], state: Optional[int],
```
This is a very general method name - is there anything more specific we could use?

kaustubh-nair replied:
I kept it general because there are multiple things happening here - returning typeahead after checking for state constraints, and setting suggestions to footer.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -251,6 +251,34 @@ def test_generic_autocomplete_emojis(self, write_box, text,
+      (['zero', 'one', 'two'], 5, None, None, False),
+      (['zero', 'one', 'two'], -5, None, None, False),
+     ], ids=[
+        'less_than_10_typeaheads',
```
minor: These are 'countable', so 'fewer' would be more accurate than 'less'

kaustubh-nair replied:
Fixed

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -251,6 +251,34 @@ def test_generic_autocomplete_emojis(self, write_box, text,
+     ], ids=[
+        'less_than_10_typeaheads',
+        'more_than_10_typeaheads',
+        'invalid_state-None',
+        'invalid_state-greater_than_possible_index',
+        'invalid_state-less_than_possible_index',
```
While accurate, are these literally invalid state? Do they represent internal state that should not happen? If we're testing the range of inputs, should we cover those that the typeahead footer understands/expects?

It's also difficult knowing whether to test this method directly, rather than just cases where it's used, which I think you're getting the feeling of from your other review comment.

kaustubh-nair replied:
@neiljp I'm not sure whose internal state you're referring to - is that zt or readline? They are invalid, in the sense that they are indexes that do not exist for the array under consideration.

We don't have to cover all cases for footer since it is tested completely in `test_ui`?

I think it is okay to test method in isolation, as long as we cover all the cases for invalid states. This would mean that other tests for autocomplete methods will require fewer test parameters, but they would be difficult to write because we'd have to use lists instead, unless of course we use really small lists.

neiljp commented on tests/ui_tools/test_boxes.py:
```diff
@@ -251,6 +251,34 @@ def test_generic_autocomplete_emojis(self, write_box, text,
+                                 is_truncated, mocker):
+        write_box.view.set_typeahead_footer = mocker.patch(
+                                'zulipterminal.ui.View.set_typeahead_footer')
+        typeaheads = ['*{}*'.format(s) for s in suggestions]
```
This is an example formatting rather than specific, right? If so, worth a comment.

kaustubh-nair replied:
Added

kaustubh-nair commented on zulipterminal/model.py:
```diff
@@ -645,6 +645,13 @@ def toggle_stream_pinned_status(self, stream_id: int) -> bool:
+            if stream['name'] == stream_name:
+                return stream_id
+
+        raise RuntimeError("Invalid stream name specified.")
```
I'm using this just to deal with the mypy error of the method not having a return value. Do I return something like `''` instead?

neiljp replied:
Well, we could do various things in case of error (empty string, `None`, custom exception, general exception), as long as we "handle" them somehow. mypy only knows return types, so exceptions act as a fix for this at some level from its perspective - and not always in a wrong way. As I said, having this exception is fine as long as we expect all stream names to be valid and otherwise it being an internal error - so here we don't expect to handle it. Later we might change it to be something that callers are expected to handle.

kaustubh-nair replied:
Okay, this should be fine for now because the recipient validity check ensures that stream names are always valid.

neiljp replied:
This was merged via the commit in #772, so this should be good for now.

kaustubh-nair commented on tests/ui_tools/test_boxes.py:
```diff
@@ -251,6 +251,38 @@ def test_generic_autocomplete_emojis(self, write_box, text,
+                       'Some general stream']),
+        ('St', 1, [['Stream 2']], ['Stream 2', 'Stream 1',
+                                   'Secret stream', 'Some general stream']),
+        pytest.param('Some gen', 1, [], ['Some general stream'],
+                     marks=pytest.mark.xfail(
+                         reason="Lacking urwid-readline support")),
```
Ohh, since this is a readline problem, we'd have to test for the actual writebox contents instead of testing the autocomplete methods.

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -109,6 +117,15 @@ def stream_box_view(self, stream_id: int, caption: str='', title: str='',
+        return self._process_typeaheads(topic_typeaheads, state,
+                                        topic_typeaheads)
+
     def _stream_box_autocomplete(self, text: str, state: Optional[int]
```
FIxed

kaustubh-nair commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -306,7 +306,12 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
+                            invalid_stream_error = (
+                                'Invalid stream name.'
+                                ' Use {} to autocomplete.'
+                                .format(keys_for_command('AUTOCOMPLETE').pop())
```
Fixed

kaustubh-nair commented on tests/ui_tools/test_boxes.py:
```diff
@@ -251,6 +251,53 @@ def test_generic_autocomplete_emojis(self, write_box, text,
+                                                    matching_streams)
+
+    @pytest.mark.parametrize('text, expected_text', [
+        ('Som', 'Some general stream'),
```
Added a normal case to make sure test works properly

kaustubh-nair commented:
@neiljp Thanks for the review! PR updated.

kaustubh-nair commented on README.md:
```diff
@@ -172,31 +172,31 @@ source ~/.zshenv
+| -----------------------------------------------------     | --------------------------------------------- |
+| Cycle through recipient and content boxes                 | <kbd>tab</kbd>                                |
+| Send a message                                            | <kbd>Alt Enter</kbd> / <kbd>Ctrl d</kbd>      |
+| Autocomplete @mentions, #stream_names, :emoji: and topics | <kbd>Ctrl</kbd> + <kbd>f</kbd>                |
```
I hope this is okay, we don't have wrapping text for topics, nor do we support typeahead for it in message box
